docs/pydoc/renderers.py

```diff
@@ -0,0 +1,112 @@
+import base64
+import dataclasses
+import io
+import os
+import sys
+import typing as t
+import warnings
+
+import docspec
+import requests
+from pydoc_markdown.contrib.renderers.markdown import MarkdownRenderer
+from pydoc_markdown.interfaces import Context, Renderer
+
+README_FRONTMATTER = """---
+title: {title}
+excerpt: {excerpt}
+category: {category}
+slug: {slug}
+order: {order}
+hidden: false
+---
+
+"""
+
+
+@dataclasses.dataclass
+class ReadmeRenderer(Renderer):
+    """
+    This custom Renderer is heavily based on the `MarkdownRenderer`,
+    it just prepends a front matter so that the output can be published
+    directly to readme.io.
+    """
+
+    # These settings will be used in the front matter output
+    title: str
+    category_slug: str
+    excerpt: str
+    slug: str
+    order: int
+    # Docs categories fetched from Readme.io
+    categories: t.Dict[str, str] = dataclasses.field(init=False)
+    # This exposes a special `markdown` settings value that can be used to pass
+    # parameters to the underlying `MarkdownRenderer`
+    markdown: MarkdownRenderer = dataclasses.field(default_factory=MarkdownRenderer)
+
+    def init(self, context: Context) -> None:
+        self.markdown.init(context)
+        version = self._doc_version()
+        self.categories = self._readme_categories(version)
+
+    def _doc_version(self) -> str:
+        """
+        Returns the docs version.
+        """
+        # full_version = about.__version__
+        # major, minor = full_version.split(".")[:2]
+        # return f"v{major}.{minor}"
+
+        # The readme.io version is hardcoded for now to manually maintain the guides
+        # within the same documentation page as the deepset Cloud Docs.
+        return "v1.0"
+
+    def _readme_categories(self, version: str) -> t.Dict[str, str]:
+        """
+        Fetch the categories of the given version from Readme.io.
+        README_API_KEY env var must be set to correctly get the categories.
+        Returns dictionary containing all the categories slugs and their ids.
+        """
+        README_API_KEY = os.getenv("README_API_KEY")
+        if not README_API_KEY:
+            warnings.warn("README_API_KEY env var is not set, using a placeholder category ID")
+            return {"open-source-sdk-methods": "ID"}
+
+        token = base64.b64encode(f"{README_API_KEY}:".encode()).decode()
```

Comment:
is the `:` after the `README_API_KEY` intended?

Reply:
Not sure tbh. :D just copied this from haystack [here](https://github.com/deepset-ai/haystack/blob/6a5fbb71189f010a2c5f24866496c347b1510818/docs/pydoc/renderers.py#L75). Seems like it's a readme implementation. It seems like readme mixed up bearer/tokens and basic authentication and expect api keys as "username" https://docs.readme.com/main/reference/intro/authentication#generating-new-api-keys- 

never seen this so far 🤷‍♂️ but works